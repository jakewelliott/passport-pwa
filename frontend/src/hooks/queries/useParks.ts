import { useQuery } from '@tanstack/react-query';
import type { Park, ParkAbbreviation } from '@/lib/mock/types';
import { api } from '@/lib/mock/api';

// ADAM:
// This is a very simple query hook.
// It's a good starting point for most queries.
// It's a good idea to have a query hook for each endpoint.

// useQuery returns an object with some useful properties:
// - data: the data returned from the query
// - isLoading: a boolean that indicates if the query is loading (data is undefined)
// - isError: a boolean that indicates if the query has an error
// - error: the error returned from the query

export const useParks = () => {
  return useQuery<Park[]>({
    queryKey: ['parks'],
    queryFn: (): Park[] => api.getParks(),
  });
};

export const usePark = (code: ParkAbbreviation) => {
  return useQuery<Park>({
    queryKey: ['park', code],
    queryFn: (): Park => api.getPark(code),
    enabled: !!code,
  });
};
