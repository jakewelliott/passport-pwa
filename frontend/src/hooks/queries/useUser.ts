import { useQuery } from '@tanstack/react-query';
import type { UserProfile } from '@/lib/mock/types';
import { api } from '@/lib/mock/api';

export const useUser = (userId?: string) => {
  // TODO: replace with auth state
  const uuid = userId ?? 'local_uuid_here';

  return useQuery<UserProfile, Error>({
    queryKey: ['user', uuid],
    // queryFn: () => api.getUserByID(uuid),
    queryFn: () => api.getUserByID(),
    enabled: !!uuid,
  });
};
