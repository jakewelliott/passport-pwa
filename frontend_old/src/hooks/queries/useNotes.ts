import { dbg } from '@/lib/debug';
import { API_NOTES_URL, fetchGet, fetchPost } from '@/lib/fetch';
import type { ParkNote } from '@/types';
import { useMutation, useQuery } from '@tanstack/react-query';

export const useNotes = () => {
    dbg('HOOK', 'useNotes');
    return useQuery<ParkNote[]>({
        queryKey: ['notes'],
        queryFn: async () => await fetchGet(API_NOTES_URL),
    });
};

export const useNote = (parkId: number) => {
    const hook = useNotes();
    const note = hook.data?.find((note) => note.parkId === parkId);
    return { ...hook, data: note };
};

export const useUpdateNote = () => {
    const { refetch } = useNotes();
    return useMutation({
        mutationFn: ({ parkId, note }: { parkId: number; note: string }) =>
            fetchPost(`${API_NOTES_URL}/${parkId}`, { note: note, updatedAt: new Date() }),
        onSuccess: () => {
            dbg('HOOK', 'useUpdateNote', 'refetching notes...');
            refetch();
        },
    });
};
