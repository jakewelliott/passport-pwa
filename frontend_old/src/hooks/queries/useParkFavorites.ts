import { dbg } from '@/lib/debug';
import { API_FAVORITE_PARKS_URL, fetchDelete, fetchGet, fetchPost } from '@/lib/fetch';
import { useMutation, useQuery } from '@tanstack/react-query';
import { toast } from 'react-toastify';

/** Gets all favorited parks */
// NOTE: only exported for testing, do not use in components!!
export const useGetFavoriteParks = () => {
    return useQuery<number[]>({
        queryKey: ['favoritedParks'],
        queryFn: () => fetchGet(API_FAVORITE_PARKS_URL),
    });
};

/** Marks a park as a favorite park for a user
 *  Call mutate with the itemId to mark the park as a favorite
 */
const useMarkParkFavorite = () => {
    const { refetch, data } = useGetFavoriteParks();

    const mutationFn = (parkId: number) => {
        dbg('MUTATE', 'toggleFavorite', { itemId: parkId });
        refetch();
        if (data?.some((x) => x === parkId)) {
            toast.error('Park is already favorited');
        }
        return fetchPost(`${API_FAVORITE_PARKS_URL}/${parkId}`, '');
    };

    return useMutation({
        mutationFn: mutationFn,
        onSuccess: () => refetch(),
    });
};

/** Removes a park from favorited parks for a user
 *  Call mutate with the itemId to remove the park from favorites
 */
const useRemoveParkFavorite = () => {
    const { refetch, data } = useGetFavoriteParks();

    const mutationFn = (parkId: number) => {
        dbg('MUTATE', 'toggleFavorite', { itemId: parkId });
        refetch();
        if (!data?.some((x) => x === parkId)) {
            toast.error('Park is not a favorite');
        }
        return fetchDelete(`${API_FAVORITE_PARKS_URL}/${parkId}`);
    };

    return useMutation({
        mutationFn: mutationFn,
        onSuccess: () => refetch(),
    });
};

export const useFavoriteParks = () => {
    const { data: allItems, isLoading: isLoadingAllItems, isFetching, refetch } = useGetFavoriteParks();
    const { mutate: markFavorite } = useMarkParkFavorite();
    const { mutate: removeFavorite } = useRemoveParkFavorite();

    const isLoading = isLoadingAllItems || !allItems;

    return {
        data: allItems,
        markFavorite,
        removeFavorite,
        isLoading,
        isFetching,
        refetch,
    };
};
