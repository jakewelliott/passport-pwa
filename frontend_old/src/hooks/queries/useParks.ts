import { API_PARKGEO_URL, API_PARKS_URL, fetchGet } from '@/lib/fetch';
import type { Park, ParkGeoData } from '@/types';
import { useQuery } from '@tanstack/react-query';

export const useParks = () => {
    const query = useQuery<Park[]>({
        queryKey: ['parks'],
        queryFn: async () => await fetchGet(API_PARKS_URL),
    });

    const parkIdToAbbreviation = (parkId: number) => {
        return query.data?.find((park) => park.id === parkId)?.abbreviation;
    };

    const parkAbbreviationToId = (parkAbbreviation: string) => {
        return query.data?.find((park) => park.abbreviation === parkAbbreviation)?.id;
    };

    return { ...query, parkIdToAbbreviation, parkAbbreviationToId };
};

export const usePark = (parkAbbreviation: string) => {
    const hook = useParks();
    const park = hook.data?.find((park) => park.abbreviation === parkAbbreviation);
    return { ...hook, data: park };
};

export const useParksGeo = () => {
    return useQuery<ParkGeoData[]>({
        queryKey: ['parkGeo'],
        queryFn: async () => await fetchGet(API_PARKGEO_URL),
    });
};
