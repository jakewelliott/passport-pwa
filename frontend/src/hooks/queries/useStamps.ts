import type { ParkAbbreviation } from '@/lib/mock/types';
import { useUser } from './useUser';
import { api } from '@/lib/mock/api';
import { useQuery } from '@tanstack/react-query';
/**
 * Gets all of the local user's stamps, empty array if user is not loaded
 * @returns The stamps for the user
 */
export const useStamps = () => {
  // re-use our query hooks whenever possible
  const { data: user } = useUser();
  return useQuery({
    queryKey: ['stamps', user?.id],
    queryFn: () => api.getUserStampsByID(user?.id || 0),
  });
};

/**
 * Gets a single stamp for a user, null if stamp not collected yet
 * @param code The code of the park to get the stamp for
 * @returns The stamp for the user
 */
export const useStamp = (code: ParkAbbreviation) => {
  // re-use our query hooks whenever possible
  const { data, isLoading } = useStamps();
  return { data: data?.find((stamp) => stamp.code === code) || null, isLoading };
};
