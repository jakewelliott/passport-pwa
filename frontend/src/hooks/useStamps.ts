import type { ParkAbbreviation } from '@/lib/mock/types';
import { useUser } from './queries/useUser';
import { api } from '@/lib/mock/api';
import { useQuery } from '@tanstack/react-query';
import { USER_LOADING_ID } from './queries/useUser';
/**
 * Gets all of the local user's stamps, empty array if user is not loaded
 * @returns The stamps for the user
 */
export const useStamps = () => {
  // re-use our query hooks whenever possible
  const { data: user } = useUser();
  // to respect rules of hooks we must always call this query hook, hence the USER_LOADING_ID
  return useQuery({
    queryKey: ['stamps', user?.id || USER_LOADING_ID],
    queryFn: () => api.getUserStampsByID(user?.id || USER_LOADING_ID),
  });
};

/**
 * Gets a single stamp for a user, null if stamp not collected yet
 * @param code The code of the park to get the stamp for
 * @returns The stamp for the user
 */
export const useStamp = (code: ParkAbbreviation) => {
  // re-use our query hooks whenever possible
  const { data, isLoading } = useStamps();
  return { data: data?.find((stamp) => stamp.code === code) || null, isLoading };
};
