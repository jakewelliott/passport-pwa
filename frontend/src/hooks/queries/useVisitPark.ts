import { useLocation } from '@/hooks/useLocation';
import { dbg } from '@/lib/debug';
import { API_VISIT_HISTORY_URL, fetchGet, fetchPost } from '@/lib/fetch';
import type { ParkVisit } from '@/types/tables';
import { useMutation, useQuery } from '@tanstack/react-query';
import { useParks } from './useParks';

/** Gets the visit history for the current user */
const useVisitsHistory = () =>
    useQuery<ParkVisit[]>({
        queryKey: ['park_visits'],
        queryFn: () => fetchGet(`${API_VISIT_HISTORY_URL}`),
    });

/** Creates a park_visit record for the current user */
const useVisitMutation = () => {
    const { geopoint } = useLocation();
    return useMutation({
        mutationFn: (parkId: number) => fetchPost(`${API_VISIT_HISTORY_URL}/${parkId}`, { geopoint: geopoint }),
        onError: (error) => {
            dbg('ERROR', 'useVisitMutation', error);
        },
    });
};

/** Marks a park as visited for today */
export const useVisitPark = () => {
    const query = useVisitsHistory();
    const mutation = useVisitMutation();
    const { data: parks } = useParks();

    const mutate = (parkId: number) => {
        dbg('MUTATE', 'useVisitPark', `visiting park ${parkId}`);

        // make sure we have the data to check
        if (query.isLoading || query.data === undefined || parks === undefined) {
            dbg('ERROR', 'useVisitPark', 'waiting on query for timecheck');
            return;
        }

        const park = parks.find((park) => park.id === parkId);

        if (!park) {
            dbg('ERROR', 'useVisitPark', `park ${parkId} not found`);
            return;
        }

        // check if we've visited the park in the last 24 hours
        const lastVisit = query.data.find((visit: ParkVisit) => visit.parkId === park.id);
        if (lastVisit && lastVisit.updatedAt > new Date(Date.now() - 24 * 60 * 60 * 1000)) {
            dbg('ERROR', 'useVisitPark', 'already visited the park in the last 24 hours');
            return;
        }

        mutation.mutate(parkId, {
            onSuccess: () => {
                query.refetch();
            },
        });
    };

    return {
        ...query,
        ...mutation,
        mutate,
    };
};
