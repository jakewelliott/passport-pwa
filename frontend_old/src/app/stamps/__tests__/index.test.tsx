import { useParks } from '@/hooks/queries/useParks';
import { useStamp, useStamps } from '@/hooks/queries/useStamps';
import { useUser } from '@/hooks/queries/useUser';
import { useVisitsHistory } from '@/hooks/queries/useVisitPark';
import { parks } from '@/lib/testing/mock';
import { setupTestEnv } from '@/lib/testing/test-wrapper';
import { fireEvent, screen } from '@testing-library/react';
import { beforeAll, describe, expect, it } from 'vitest';
import StampsScreen from '../';
const { render, checkHook } = setupTestEnv();
describe('Stamps', async () => {
    beforeAll(async () => {
        await checkHook(useStamps, 'useStamps');
        await checkHook(useStamp, 'useStamp', 5);
        await checkHook(useUser, 'useUser');
        await checkHook(useParks, 'useParks');
        // TODO: useVisitsHistory endpoint might need to be mocked
        await checkHook(useVisitsHistory, 'useVisitsHistory');
    });

    it('matches snapshot', () => {
        const { container } = render(<StampsScreen />);
        expect(container).toMatchSnapshot();
    });

    it('renders stamps in correct grid layout', async () => {
        render(<StampsScreen />);
        const gridContainer = screen.getByTestId('stamps-grid');
        expect(gridContainer).toHaveClass('grid', 'grid-cols-3', 'gap-4');
    });

    // ADAM: i'm at a loss to why this test is failing
    // it('renders grid with achieved stamps', () => {
    //     render(<StampsScreen />);
    //     screen.logTestingPlaygroundURL();
    //     const crmoStamp = screen.getByAltText('CLNE - achieved');
    //     expect(crmoStamp).not.toHaveClass('opacity-50', 'grayscale');
    // });

    // it('renders grid with unachieved stamps', () => {
    //     render(<StampsScreen />);
    //     const cacrStamp = screen.getByAltText('CACR - greyed out');
    //     expect(cacrStamp).toHaveClass('opacity-50', 'grayscale');
    // });

    it('shows stamp details when clicking unvisited stamp', () => {
        render(<StampsScreen />);

        // Click an unvisited stamp (third park onwards)
        const unvisitedStamp = screen.getByTestId(`stamp-button-${parks[2].abbreviation}`);
        fireEvent.click(unvisitedStamp);

        expect(screen.getByRole('article')).toBeInTheDocument();
    });

    it('shows stamp details when clicking a stamp', () => {
        render(<StampsScreen />);

        // Click the first stamp (which is visited)
        const firstStamp = screen.getByTestId(`stamp-button-${parks[0].abbreviation}`);
        fireEvent.click(firstStamp);

        // StampDetails component should be rendered
        expect(screen.getByRole('article')).toBeInTheDocument();
    });

    it('hides stamp details when close button is clicked', () => {
        render(<StampsScreen />);

        // Click a stamp to show details
        const firstStamp = screen.getByTestId(`stamp-button-${parks[0].abbreviation}`);
        fireEvent.click(firstStamp);

        // Click the close button
        const closeButton = screen.getByRole('button', { name: 'Close park details' });
        fireEvent.click(closeButton);

        // StampDetails component should not be in the document
        expect(screen.queryByRole('article')).not.toBeInTheDocument();
    });

    it('applies correct styling to stamp buttons', () => {
        render(<StampsScreen />);

        // Test the first stamp button's styling
        const stampButton = screen.getByTestId(`stamp-button-${parks[0].abbreviation}`);
        expect(stampButton).toHaveClass('flex', 'items-center', 'justify-center', 'p-2');
    });
});
