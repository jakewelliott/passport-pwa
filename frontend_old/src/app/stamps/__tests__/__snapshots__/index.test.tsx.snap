// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stamps > matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="relative flex gap-2"
    >
      <div
        class="relative flex-1"
      >
        <input
          class="w-full rounded-lg border border-system_gray p-3 pr-7 focus:border-secondary_darkteal focus:outline-none focus:ring-1 focus:ring-secondary_darkteal focus:ring-opacity-100"
          placeholder="Search parks..."
          type="text"
          value=""
        />
      </div>
      <button
        class="w-auto rounded-lg border border-system_gray bg-system_white px-3 py-2 transition-all duration-200"
        style="display: inline;"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
          />
        </svg>
      </button>
    </div>
    <div
      class="grid grid-cols-3 gap-4"
      data-testid="stamps-grid"
    >
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-CABE"
        type="button"
      >
        <img
          alt="CABE - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-CABE"
          src="/stamps/CABE.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-CACR"
        type="button"
      >
        <img
          alt="CACR - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-CACR"
          src="/stamps/CACR.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-CHRO"
        type="button"
      >
        <img
          alt="CHRO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-CHRO"
          src="/stamps/CHRO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-CLNE"
        type="button"
      >
        <img
          alt="CLNE - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-CLNE"
          src="/stamps/CLNE.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-CRMO"
        type="button"
      >
        <img
          alt="CRMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-CRMO"
          src="/stamps/CRMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-DISW"
        type="button"
      >
        <img
          alt="DISW - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-DISW"
          src="/stamps/DISW.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-ELKN"
        type="button"
      >
        <img
          alt="ELKN - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-ELKN"
          src="/stamps/ELKN.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-ENRI"
        type="button"
      >
        <img
          alt="ENRI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-ENRI"
          src="/stamps/ENRI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-FALA"
        type="button"
      >
        <img
          alt="FALA - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-FALA"
          src="/stamps/FALA.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-FOFI"
        type="button"
      >
        <img
          alt="FOFI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-FOFI"
          src="/stamps/FOFI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-FOMA"
        type="button"
      >
        <img
          alt="FOMA - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-FOMA"
          src="/stamps/FOMA.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-GOCR"
        type="button"
      >
        <img
          alt="GOCR - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-GOCR"
          src="/stamps/GOCR.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-GORG"
        type="button"
      >
        <img
          alt="GORG - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-GORG"
          src="/stamps/GORG.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-GRMO"
        type="button"
      >
        <img
          alt="GRMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-GRMO"
          src="/stamps/GRMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-HABE"
        type="button"
      >
        <img
          alt="HABE - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-HABE"
          src="/stamps/HABE.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-HARO"
        type="button"
      >
        <img
          alt="HARO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-HARO"
          src="/stamps/HARO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-HARI"
        type="button"
      >
        <img
          alt="HARI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-HARI"
          src="/stamps/HARI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-JORI"
        type="button"
      >
        <img
          alt="JORI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-JORI"
          src="/stamps/JORI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-JONE"
        type="button"
      >
        <img
          alt="JONE - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-JONE"
          src="/stamps/JONE.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-JORD"
        type="button"
      >
        <img
          alt="JORD - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-JORD"
          src="/stamps/JORD.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-KELA"
        type="button"
      >
        <img
          alt="KELA - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-KELA"
          src="/stamps/KELA.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-LAJA"
        type="button"
      >
        <img
          alt="LAJA - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-LAJA"
          src="/stamps/LAJA.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-LANO"
        type="button"
      >
        <img
          alt="LANO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-LANO"
          src="/stamps/LANO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-LAWA"
        type="button"
      >
        <img
          alt="LAWA - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-LAWA"
          src="/stamps/LAWA.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-LURI"
        type="button"
      >
        <img
          alt="LURI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-LURI"
          src="/stamps/LURI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-MARI"
        type="button"
      >
        <img
          alt="MARI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-MARI"
          src="/stamps/MARI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-MEMO"
        type="button"
      >
        <img
          alt="MEMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-MEMO"
          src="/stamps/MEMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-MEMI"
        type="button"
      >
        <img
          alt="MEMI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-MEMI"
          src="/stamps/MEMI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-MOMO"
        type="button"
      >
        <img
          alt="MOMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-MOMO"
          src="/stamps/MOMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-MOJE"
        type="button"
      >
        <img
          alt="MOJE - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-MOJE"
          src="/stamps/MOJE.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-MOMI"
        type="button"
      >
        <img
          alt="MOMI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-MOMI"
          src="/stamps/MOMI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-NERI"
        type="button"
      >
        <img
          alt="NERI - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-NERI"
          src="/stamps/NERI.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-OCMO"
        type="button"
      >
        <img
          alt="OCMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-OCMO"
          src="/stamps/OCMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-PETT"
        type="button"
      >
        <img
          alt="PETT - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-PETT"
          src="/stamps/PETT.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-PIMO"
        type="button"
      >
        <img
          alt="PIMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-PIMO"
          src="/stamps/PIMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-RARO"
        type="button"
      >
        <img
          alt="RARO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-RARO"
          src="/stamps/RARO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-REMO"
        type="button"
      >
        <img
          alt="REMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-REMO"
          src="/stamps/REMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-SILA"
        type="button"
      >
        <img
          alt="SILA - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-SILA"
          src="/stamps/SILA.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-SOMO"
        type="button"
      >
        <img
          alt="SOMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-SOMO"
          src="/stamps/SOMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-STMO"
        type="button"
      >
        <img
          alt="STMO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-STMO"
          src="/stamps/STMO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-WEWO"
        type="button"
      >
        <img
          alt="WEWO - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-WEWO"
          src="/stamps/WEWO.svg"
        />
      </button>
      <button
        class="flex items-center justify-center p-2"
        data-testid="stamp-button-WIUM"
        type="button"
      >
        <img
          alt="WIUM - greyed out"
          class="opacity-50 grayscale"
          data-testid="stamp-image-WIUM"
          src="/stamps/WIUM.svg"
        />
      </button>
    </div>
  </div>
</div>
`;
